geniza/corpus/management/commands/add_fragment_urls.py

```diff
@@ -0,0 +1,144 @@
+import csv
+import re
+from collections import defaultdict
+
+from django.conf import settings
+from django.contrib.admin.models import CHANGE, LogEntry
+from django.contrib.auth.models import User
+from django.contrib.contenttypes.models import ContentType
+from django.core.management.base import BaseCommand, CommandError
+from parasolr.django.signals import IndexableSignalHandler
+
+from geniza.corpus.models import Fragment
+
+
+class Command(BaseCommand):
+    """Takes a CSV of shelfmarks and view URLs and/or IIIF URLs, update
+    corresponding Fragment records in the database with those URLs.
+    Expects CSV headers 'shelfmark' and one or both of 'url' and 'iiif_url'"""
+
+    help = __doc__
+
+    def __init__(self, *args, **options):
+        super().__init__(*args, **options)
+
+        self.stats = defaultdict(int)
```

Comment:
obvi not a blocker for merging, but for stat collection things like this I've used [`Counter`](https://docs.python.org/3/library/collections.html#collections.Counter), which behaves the same (counts of nonexistent elements are zero instead of `KeyError`) and I think might have a few optimizations under the hood

Reply:
ah, thanks — I learned about this recently but I forget